### PR TITLE
[Doc] Document LossModule and collector weight sharing

### LINKED ISSUE #3032: [Feature Request] Documentation about weight update propagation
First of all, thank you so much to everyone working on this project. I'm currently using it in several projects and it's (mostly) been a very pleasant experience.

But there is one thing in torchrl that I find particularly confusing.

## Motivation

I'm confused about whether weight updates in `LossModule`s get propagated back to the original module that was passed in to the `LossModule`. Let's take `DQNLoss` as an example and suppose we have a `TensorDictModule` called `module` that we pass in as the `value_network` argument. As I understand it, the loss module has its own weight copies of the network that was passed in as `value_network`. So when we call `loss_module.backward()`, the weights of the original network are **not** updated.

When we give the original module (or rather an epsilon-greedy version of it) to a `SyncDataCollector`, we have to call `collector.update_policy_weights_()` in order for the loss module weights to get copied to the original module. But what if we want to do inference without a collector?

I've seen snippets like
```python
with self.loss_module.qvalue_network_params["module"]["0"].to_module(self.module):
  # self.module can now be used with updated weights
```
but I'm not sure if that's the proper way to do it...

I've tried looking into the source code for `SyncDataCollector` and `VanillaWeightUpdater` but I think that this would be a very useful piece of documentation to have.

## Solution

A documentation page that briefly mentions
- whether weight updates from the loss module are propagated back to the original module
- an example of how to use a policy for inference (without using a collector)

## Alternatives

Maybe I'm not understanding something that is obvious to everyone else. In that case, please point me to a source that explains these things.

## Checklist

- [X] I have checked that there is no similar issue in the repo (**required**)

### PR DESCRIPTION
**Pull request by YeonwooSung**
## Description

Document how `LossModule` and collectors share policy weights, and when `collector.update_policy_weights_()` is actually required.

The objectives landing page (and the DDPG tutorial) said losses "contain a copy of the trainable parameters". That wording made it look like an optimizer step on the loss would not update the original module, and that a collector always needed an explicit sync. Maintainer answers on #3032 already settled the semantics; this PR writes them down.

- Passing a module into a `LossModule` does **not** copy it. The same parameters are used in-place; optimizer steps on the loss update the original module.
- A CPU policy passed to a collector is moved to shared memory in-place. A CUDA policy is already shared.
- `update_policy_weights_()` is **required** when `policy_device` or `device` remaps the policy (a `.to(...)` copy is made). Calling it anyway is good practice but not load-bearing unless devices were remapped.

Cross-links: `LossModule` / objectives landing page ↔ collectors weight-sync page.

### Code example

An optimizer registered on the loss updates the original actor because their trainable parameters share storage.

```python
import torch
from torch import nn
from torchrl.modules import Actor, ValueOperator
from torchrl.objectives import DDPGLoss

actor = Actor(nn.Linear(3, 1))
critic = ValueOperator(nn.Linear(4, 1), in_keys=["observation", "action"])
loss = DDPGLoss(actor, critic, delay_actor=False, delay_value=False)
assert actor is loss.actor_network
optimizer = torch.optim.SGD(loss.parameters(), lr=0.1)
weight = next(actor.parameters())
before = weight.detach().clone()
weight.square().sum().backward()  # Small objective to demonstrate the storage link.
optimizer.step()
assert not torch.equal(weight.detach(), before)
# A collector using copied policy storage additionally needs update_policy_weights_().
```

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [x] I have raised an issue to propose this change ([required](https://github.com/pytorch/rl/issues) for new features and bug fixes)

close #3032

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Review by vmoens — APPROVED** [review]
Verified the parameter-sharing implementation and collector copy conditions, and executed the new weight-sharing example. No actionable findings.

**Comment by YeonwooSung** [comment]
Rebased onto current `main`. The weight-sharing notes still match the LossModule and collector copy conditions after the collector progress / replay-write and trainer-resume work that landed since this PR.

Re-ran the DDPG identity / shared-storage / optimizer-step example after the rebase.

The previous collectors / gym / GPU CI failures were on a stale merge base and are outside this documentation diff.

File: docs/source/reference/collectors_weightsync.rst [hunk 1]
```diff
@@ -3,6 +3,8 @@
 Weight Synchronization
 ======================
 
+.. _ref_collectors_weightsync:
```
Comment (vmoens) [inline docs/source/reference/collectors_weightsync.rst]: [P2] Place the reference target before the heading

This target follows the `Weight Synchronization` heading, so Sphinx has no title/caption for the new bare `:ref:` links elsewhere in the PR. A minimal Sphinx 8.1.3 build using this unchanged page opening emits `Failed to create a cross reference. A title or caption not found: 'ref_collectors_weightsync' [ref.ref]` and fails with `-W`. Moving the target immediately before the heading makes the same build pass.

File: docs/source/reference/objectives.rst [hunk 2]
```diff
@@ -39,6 +40,55 @@ Quick Example
...
     total_loss = sum(v for k, v in loss_vals.items() if k.startswith("loss_"))
 
+.. _ref_lossmodule_weight_sharing:
+
+Weight sharing
+--------------
+
+Passing a module into a :class:`~torchrl.objectives.LossModule` does **not**
+copy it. The loss stores the same module and the same parameter tensors; an
+optimizer step on ``loss.parameters()`` updates the original network in-place.
```
Comment (vmoens) [inline docs/source/reference/objectives.rst]: [P2] Exempt expanded ensembles from the parameter-sharing guarantee

This guarantee is broader than `LossModule` supports. `convert_to_functional(..., expand_dim=N)` clones/resamples independent ensemble parameters, as used by default `SACLoss` for its critics. In a live check, `loss.qvalue_network is critic` was true, but the original critic weight had shape `(1, 4)` while the optimized parameter had shape `(2, 1, 4)`; an optimizer step changed the ensemble and left the original critic unchanged. Calling the original critic can therefore use untrained parameters despite object identity. Please scope the guarantee to the non-expanded shared parameters demonstrated by the DDPG/actor example, and distinguish expanded/stacked ensembles and delayed targets. Apply the same qualification to the new wording in `LossModule` and `convert_to_functional` docstrings.

**Review by vmoens — CHANGES_REQUESTED** [review]
Rechecked the latest head `95b21b4511faab4bed6b2fd2949b6334f305edba`. The earlier device-argument/storage-ownership issue is fixed, and the DDPG sharing example passes. Two remaining issues are reproduced: the new weight-sync reference has no section caption, and the general no-copy claim does not hold for expanded critic ensembles.

**Comment by YeonwooSung** [comment]
Addressed the two review notes on `95b21b451` and rebased onto current `main`.

- `.. _ref_collectors_weightsync:` now sits immediately before the Weight Synchronization heading so the `:ref:` links have a title.
- The no-copy guarantee is limited to non-expanded parameters (DDPG actor). Expanded ensembles (`expand_dim`, default SAC critics) and delayed targets are called out in the Weight sharing section, the `LossModule` docstring, and `convert_to_functional`.

File: docs/source/reference/collectors_weightsync.rst (hunk 1 above)
Comment (YeonwooSung) [reply docs/source/reference/collectors_weightsync.rst]: Moved `.. _ref_collectors_weightsync:` to immediately before the Weight Synchronization heading so the new `:ref:` links pick up that title.

File: docs/source/reference/objectives.rst (hunk 2 above)
Comment (YeonwooSung) [reply docs/source/reference/objectives.rst]: Scoped the no-copy claim to non-expanded parameters, matching the DDPG actor example. The Weight sharing section, `LossModule` docstring, and `convert_to_functional` now distinguish expanded/stacked ensembles (`expand_dim`, default SAC critics) and delayed targets (`create_target_params=True`). Reproduced the SAC critic case: `loss.qvalue_network is critic` holds, the original weights stay unexpanded and untrained, and the optimizer updates the ensemble.

File: docs/source/reference/collectors_weightsync.rst [hunk 3]
```diff
@@ -64,6 +66,32 @@ Each of these classes is detailed below.
...
+Passing ``policy_device`` or ``device`` is not by itself evidence of a copy. If
+the policy is already on the requested device, the collector can keep the same
+object and storage. If a move is needed, it creates a separate policy and
+:meth:`~torchrl.collectors.Collector.update_policy_weights_` must propagate the
+trained weights.
+
+Calling :meth:`~torchrl.collectors.Collector.update_policy_weights_` anyway
+is conservative good practice. Use it to push weights through a
+:class:`~torchrl.weight_update.WeightSyncScheme` or another configured transport
+whenever the policies do not share storage (the rest of this page).
```
Comment (vmoens) [inline docs/source/reference/collectors_weightsync.rst:93]: I think we should have a bunch of example blocks here, that'd be much clearer.

Comment (vmoens) [reply docs/source/reference/collectors_weightsync.rst:93]: these should include various things, like using the device kwargs in collector, using remote collectors, using policy factories, etc.

Comment (YeonwooSung) [reply docs/source/reference/collectors_weightsync.rst:93]: Added example blocks for the cases that decide whether the call is load-bearing: shared storage on a direct collector, same-device `policy_device`/`device` (no copy), a device kwargs copy, local process workers, `policy_factory` workers, and a remote `backend="ray"` collector (`rpc`/`distributed` follow the same pattern).

**Comment by YeonwooSung** [comment]
Addressed the request for example blocks in the `update_policy_weights_` section and rebased onto current `main`.

The new snippets cover shared storage, same-device vs copying `policy_device`/`device`, local process workers, `policy_factory`, and a remote Ray collector.
